orttraining/orttraining/training_api/module.cc

```diff
@@ -103,39 +105,45 @@ Module::Module(const std::string& train_model_path_or_bytes,
       ORT_ENFORCE(target_device == *(it->device), "Inconsistent device requirements found for input: ", param_name);
     }
 
-    // Create parameter value copy with corresponding device user sets the session on.
-    // We did not re-use the data even CPU tensor is needed.
     // TODO(pengwa): consider whether we should alloc contiguous buffer for parameters or gradients.
-    OrtValue target_ortvalue;
-    auto allocator = train_sess_state.GetAllocator(target_device);
-    ORT_ENFORCE(allocator != nullptr);
-
-    Tensor::InitOrtValue(source_tensor.DataType(),
-                         source_tensor.Shape(),
-                         allocator, target_ortvalue);
-    Tensor* target_tensor_ptr = target_ortvalue.GetMutable<Tensor>();
-    ORT_THROW_IF_ERROR(train_sess_state.GetDataTransferMgr().CopyTensor(source_tensor, *target_tensor_ptr));
+    // Copy ortvalue buffer from CPU to target_device for this "param_name" (based on graph partitioning)
+    // Only copies data if target device is not the same as the current device the buffer is placed on
+    auto params_iter = named_parameters_.find(param_name);
+    ORT_ENFORCE(params_iter != named_parameters_.end());
+    OrtValue& param_data = params_iter->second->Data();
+
+    ORT_ENFORCE(param_data.IsTensor());
+    const Tensor& param_data_tensor = param_data.Get<Tensor>();
+    // if the source device type is already same as target device skip copy
+    if (param_data_tensor.Location().device.Type() != target_device.Type()) {
+      // TODO: move this outside of the for loop?
+      auto target_allocator = train_sess_state.GetAllocator(target_device);
+      ORT_ENFORCE(target_allocator != nullptr);
+
+      // create a new tensor on the target_device and switch the source_ortvalue to point to this new tensor
+      auto target_tensor = std::make_unique<Tensor>(param_data_tensor.DataType(), param_data_tensor.Shape(), target_allocator);
+      ORT_THROW_IF_ERROR(train_sess_state.GetDataTransferMgr().CopyTensor(param_data_tensor, *target_tensor.get()));
+      auto ml_tensor = DataTypeImpl::GetType<Tensor>();
+      param_data.Init(target_tensor.release(), ml_tensor, ml_tensor->GetDeleteFunc());
```

Comment:
does this release the original buffer on CPU?

Reply:
It should because only ORTValue has the shared pointer for the buffer... Adding a todo to test this.